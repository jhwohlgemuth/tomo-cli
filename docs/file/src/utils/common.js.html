<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <base data-ice="baseUrl" href="../../../">
  <title data-ice="title">src/utils/common.js | tomo-cli</title>
  <link type="text/css" rel="stylesheet" href="css/style.css">
  <link type="text/css" rel="stylesheet" href="css/prettify-tomorrow.css">
  <script src="script/prettify/prettify.js"></script>
  <script src="script/manual.js"></script>
<meta name="description" content="A friendly command line tool designed to help create sustainable software using web technology"><meta property="twitter:card" content="summary"><meta property="twitter:title" content="tomo-cli"><meta property="twitter:description" content="A friendly command line tool designed to help create sustainable software using web technology"></head>
<body class="layout-container" data-ice="rootContainer">

<header>
  <a href="./">Home</a>
  
  <a href="identifiers.html">Reference</a>
  <a href="source.html">Source</a>
  <a href="test.html" data-ice="testLink">Test</a>
  <div class="search-box">
  <span>
    <img src="./image/search.png">
    <span class="search-input-edge"></span><input class="search-input"><span class="search-input-edge"></span>
  </span>
    <ul class="search-result"></ul>
  </div>
<a style="position:relative; top:3px;" href="https://github.com/jhwohlgemuth/tomo-cli"><img width="20px" src="./image/github.png"></a></header>

<nav class="navigation" data-ice="nav"><div>
  <ul>
    
  <li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-CommandError">CommandError</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-OfflineWarning">OfflineWarning</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-Task">Task</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-TaskList">TaskList</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-UI">UI</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-Warning">Warning</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-populateQueue">populateQueue</a></span></span></li>
<li data-ice="doc"><a data-ice="dirPath" class="nav-dir-path" href="identifiers.html#utils">utils</a><span data-ice="kind" class="kind-class">C</span><span data-ice="name"><span><a href="class/src/utils/BasicEditor.js~BasicEditor.html">BasicEditor</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-class">C</span><span data-ice="name"><span><a href="class/src/utils/MakefileEditor.js~MakefileEditor.html">MakefileEditor</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-class">C</span><span data-ice="name"><span><a href="class/src/utils/Scaffolder.js~Scaffolder.html">Scaffolder</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-allDoExist">allDoExist</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-allDoExistSync">allDoExistSync</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-allDoNotExist">allDoNotExist</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-allDoNotExistSync">allDoNotExistSync</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-format">format</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-getBinDirectory">getBinDirectory</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-getCommandDirectory">getCommandDirectory</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-parse">parse</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-someDoExist">someDoExist</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-someDoExistSync">someDoExistSync</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-createJsonEditor">createJsonEditor</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-createModuleEditor">createModuleEditor</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-getIntendedInput">getIntendedInput</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-getVersions">getVersions</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-install">install</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-function">F</span><span data-ice="name"><span><a href="function/index.html#static-function-verifyRustInstallation">verifyRustInstallation</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-variable">V</span><span data-ice="name"><span><a href="variable/index.html#static-variable-BabelConfigModuleEditor">BabelConfigModuleEditor</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-variable">V</span><span data-ice="name"><span><a href="variable/index.html#static-variable-EslintConfigModuleEditor">EslintConfigModuleEditor</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-variable">V</span><span data-ice="name"><span><a href="variable/index.html#static-variable-PackageJsonEditor">PackageJsonEditor</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-variable">V</span><span data-ice="name"><span><a href="variable/index.html#static-variable-PostcssConfigEditor">PostcssConfigEditor</a></span></span></li>
<li data-ice="doc"><span data-ice="kind" class="kind-variable">V</span><span data-ice="name"><span><a href="variable/index.html#static-variable-WebpackConfigEditor">WebpackConfigEditor</a></span></span></li>
</ul>
</div>
</nav>

<div class="content" data-ice="content"><h1 data-ice="title">src/utils/common.js</h1>
<pre class="source-code line-number raw-source-code"><code class="prettyprint linenums" data-ice="content">import {join} from &apos;path&apos;;
import {pathExists, pathExistsSync} from &apos;fs-extra&apos;;
import {which} from &apos;shelljs&apos;;
import {isNull, negate} from &apos;lodash&apos;;
import prettier from &apos;prettier&apos;;

const PRETTIER_OPTIONS = {
    bracketSpacing: false,
    parser: &apos;json-stringify&apos;,
    printWidth: 80,
    tabWidth: 4,
    quotes: true
};
export const parse = data =&gt; JSON.parse(JSON.stringify(data));
export const getCommandDirectory = command =&gt; {
    const data = which(command);
    const commandExists = negate(isNull)(data);
    return commandExists ? data.toString().split(command)[0] : &apos;&apos;;
};
export const getBinDirectory = path =&gt; {
    const [packageDirectory] = path.split(&apos;Makefile&apos;);
    return `${packageDirectory}node_modules/.bin/`;
};
/**
 * Format input code using Prettier
 * @param {*} [code=&apos;&apos;] Code to be formatted
 * @example &lt;caption&gt;Prettier options&lt;/caption&gt;
 * {
 *     bracketSpacing: false,
 *     parser: &apos;json-stringify&apos;,
 *     printWidth: 80,
 *     tabWidth: 4,
 *     quotes: true
 * }
 * @return {string} Code formatted by Prettier
 */
export const format = (code = {}) =&gt; prettier.format(JSON.stringify(code), PRETTIER_OPTIONS).replace(/&quot;/g, &apos;&apos;);
/**
 * Check that at least one file or files exist
 * @param  {...string} args File or folder path(s)
 * @example
 * // some/folder/
 * //   &#x251C;&#x2500; foo.js
 * //   &#x2514;&#x2500;&#x2500; bar.js
 * const hasFoo = someDoExist(&apos;some/folder/foo.js&apos;);
 * const hasBaz = someDoExist(&apos;some/folder/baz.js&apos;);
 * const hasSomething = someDoExist(&apos;some/folder/bar.js&apos;, &apos;some/folder/baz.js&apos;);
 * console.log(hasFoo); // true
 * console.log(hasBaz); // false
 * console.log(hasSomething); // true
 * @return {boolean} Some files/path do exist (true) or all do not exist (false)
 */
export const someDoExist = async (...args) =&gt; {
    const checks = await Promise.all(args.map(val =&gt; pathExists(join(process.cwd(), val))));
    return checks.some(Boolean);
};
/**
 * Check that at least one file or files exist (synchronous version of {@link someDoExist})
 * @param  {...string} args File or folder path(s)
 * @return {boolean} Some files/path do exist (true) or all do not exist (false)
 */
export const someDoExistSync = (...args) =&gt; args.map(val =&gt; pathExistsSync(join(process.cwd(), val))).some(Boolean);
/**
 * Check that all files exist
 * @param  {...string} args File of folder paths
 * @return {boolean} All files/paths exist (true) or do not (false)
 */
export const allDoExist = async (...args) =&gt; {
    const checks = await Promise.all(args.map(val =&gt; pathExists(join(process.cwd(), val))));
    return checks.every(Boolean);
};
/**
 * Check that all files exist (synchronous version of {@link allDoExist})
 * @param  {...string} args File of folder paths
 * @return {boolean} All files/paths exist (true) or do not (false)
 */
export const allDoExistSync = (...args) =&gt; args.map(val =&gt; pathExistsSync(join(process.cwd(), val))).every(Boolean);
/**
 * Check that all files do not exist
 * @example
 * // some/folder/
 * //   &#x251C;&#x2500; foo.js
 * //   &#x2514;&#x2500;&#x2500; bar.js
 * const noPackageJson = allDoNotExist(&apos;some/folder/package.json&apos;);
 * console.log(noPackageJson); // true
 * @param  {...string} args File or folder path(s)
 * @return {boolean} All files/paths do not exist (true) or some do (false)
 */
export const allDoNotExist = async (...args) =&gt; {
    const checks = await Promise.all(args.map(val =&gt; pathExists(join(process.cwd(), val))));
    return checks.every(val =&gt; !val);
};
/**
 * Check that all files do not exist (synchronous version of {@link allDoNotExist})
 * @param  {...string} args File or folder path(s)
 * @return {boolean} All files/paths do not exist (true) or some do (false)
 */
export const allDoNotExistSync = (...args) =&gt; args.map(val =&gt; pathExistsSync(join(process.cwd(), val))).every(val =&gt; !val);</code></pre>

</div>

<footer class="footer">
  Generated by <a href="https://esdoc.org">ESDoc<span data-ice="esdocVersion">(1.1.0)</span><img src="./image/esdoc-logo-mini-black.png"></a>
</footer>

<script src="script/search_index.js"></script>
<script src="script/search.js"></script>
<script src="script/pretty-print.js"></script>
<script src="script/inherited-summary.js"></script>
<script src="script/test-summary.js"></script>
<script src="script/inner-link.js"></script>
<script src="script/patch-for-local.js"></script>
</body>
</html>
