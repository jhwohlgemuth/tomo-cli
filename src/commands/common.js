import {join} from 'path';
import {mkdirp} from 'fs-extra';
import {
    PackageJsonEditor,
    Scaffolder,
    allDoNotExist
} from '../api';

/** @ignore */
export const createEditorConfiguration = [
    {
        text: 'Create editor configuration file',
        task: async ({overwrite}) => {
            await (new Scaffolder(join(__dirname, 'common', 'templates')))
                .overwrite(overwrite)
                .target('.')
                .copy('.editorconfig', '.editorconfig')
                .commit();
        },
        condition: () => allDoNotExist('.editorconfig')
    }
];
/** @ignore */
export const createPackageJson = [
    {
        text: 'Create package.json',
        task: async () => {
            await (new PackageJsonEditor())
                .create()
                .commit();
        },
        condition: () => allDoNotExist('package.json')
    }
];
/** @ignore */
export const createSourceDirectory = [
    {
        text: 'Create source directory',
        task: ({sourceDirectory}) => mkdirp(sourceDirectory),
        condition: ({sourceDirectory}) => allDoNotExist(sourceDirectory)
    }
];