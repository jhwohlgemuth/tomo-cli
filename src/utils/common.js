import {join} from 'path';
import {is} from 'ramda';
import {pathExists, pathExistsSync} from 'fs-extra';
import prettier from 'prettier';

const PRETTIER_OPTIONS = {
    bracketSpacing: false,
    parser: 'json-stringify',
    printWidth: 150,
    tabWidth: 4,
    singleQuote: true
};
const newMap = val => new Map(val);
const joinPath = name => join(process.cwd(), name);
const checkPathExists = name => name |> joinPath |> pathExists;
const checkPathExistsSync = name => name |> joinPath |> pathExistsSync;
export const dict = val => val |> Object.entries |> newMap;
export const isEmptyString = data => typeof data === 'string' && data.length === 0;
export const parse = data => data |> JSON.stringify |> JSON.parse;
export const maybeApply = (val, options) => is(Function)(val) ? val(options) : val;
export const getBinDirectory = path => {
    const [packageDirectory] = path.split('Makefile');
    return `${packageDirectory}node_modules/.bin/`;
};
/**
 * Format input code using Prettier
 * @param {*} [code=''] Code to be formatted
 * @example <caption>Prettier options</caption>
 * {
 *     bracketSpacing: false,
 *     parser: 'json-stringify',
 *     printWidth: 150,
 *     tabWidth: 4,
 *     singleQuote: true
 * }
 * @return {string} Code formatted by Prettier
 */
export const format = (code = {}) => prettier
    .format(JSON.stringify(code), PRETTIER_OPTIONS)
    .replace(/"/g, '');
/**
 * Check that at least one file or files exist
 * @param  {...string} args File or folder path(s)
 * @example
 * // some/folder/
 * //   ├─ foo.js
 * //   └── bar.js
 * const hasFoo = someDoExist('some/folder/foo.js');
 * const hasBaz = someDoExist('some/folder/baz.js');
 * const hasSomething = someDoExist('some/folder/bar.js', 'some/folder/baz.js');
 * console.log(hasFoo); // true
 * console.log(hasBaz); // false
 * console.log(hasSomething); // true
 * @return {boolean} Some files/path do exist (true) or all do not exist (false)
 */
export const someDoExist = async (...args) => {
    const checks = await Promise.all(args.map(checkPathExists));
    return checks.some(Boolean);
};
/**
 * Check that at least one file or files exist (synchronous version of {@link someDoExist})
 * @param  {...string} args File or folder path(s)
 * @return {boolean} Some files/path do exist (true) or all do not exist (false)
 */
export const someDoExistSync = (...args) => args.map(checkPathExistsSync).some(Boolean);
/**
 * Check that all files exist
 * @param  {...string} args File of folder paths
 * @return {boolean} All files/paths exist (true) or do not (false)
 */
export const allDoExist = async (...args) => {
    const checks = await Promise.all(args.map(checkPathExists));
    return checks.every(Boolean);
};
/**
 * Check that all files exist (synchronous version of {@link allDoExist})
 * @param  {...string} args File of folder paths
 * @return {boolean} All files/paths exist (true) or do not (false)
 */
export const allDoExistSync = (...args) => args.map(checkPathExistsSync).every(Boolean);
/**
 * Check that all files do not exist
 * @example
 * // some/folder/
 * //   ├─ foo.js
 * //   └── bar.js
 * const noPackageJson = allDoNotExist('some/folder/package.json');
 * console.log(noPackageJson); // true
 * @param  {...string} args File or folder path(s)
 * @return {boolean} All files/paths do not exist (true) or some do (false)
 */
export const allDoNotExist = async (...args) => {
    const checks = await Promise.all(args.map(checkPathExists));
    return checks.every(val => !val);
};
/**
 * Check that all files do not exist (synchronous version of {@link allDoNotExist})
 * @param  {...string} args File or folder path(s)
 * @return {boolean} All files/paths do not exist (true) or some do (false)
 */
export const allDoNotExistSync = (...args) => args.map(checkPathExistsSync).every(val => !val);