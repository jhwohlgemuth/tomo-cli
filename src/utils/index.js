import execa from 'execa';
import semver from 'semver';
import {complement, has, head} from 'ramda';
import isOnline from 'is-online';
import {oneLineTrim} from 'common-tags';
import validate from 'validate-npm-package-name';
import {findBestMatch} from 'string-similarity';
import {dict} from './common';
import createJsonEditor from './createJsonEditor';
import {createFunctionModuleEditor, createModuleEditor} from './createModuleEditor';

const {assign, keys} = Object;
const {isArray} = Array;
export const isUniqueTask = ({text}, index, tasks) => tasks.map(({text}) => text).indexOf(text) === index;
export const isValidTask = val => has('text', val) && has('task', val) && (typeof val.text === 'string') && (typeof val.task === 'function');
export const withOptions = val => options => ({...options, ...val});
/**
 * Choose tasks based on CLI options
 * @param {Object} choices Object to create choice dictionary from
 * @return {function} Accepts CLI options and returns array of tasks
 */
export const choose = choices => options => {
    const possible = keys(choices);
    const passed = keys(options);
    const lookup = dict(choices);
    const DEFAULT = lookup.has('default') ? lookup.get('default') : lookup.get(possible[0]);
    const [choice] = possible.filter(val => passed.includes(val)).filter(val => options[val]);
    return choice ? lookup.get(choice) : DEFAULT;
};
/**
 * Use string-similarity module to determine closest matching string
 * @param {Object} commands Object with commands as key values, terms as key values for each command object
 * @param {string} command Command string input
 * @param {string[]} [terms=[]] Terms input
 * @example
 * const [intendedCommand, intendedTerms] = getIntendedInput(commands, command, terms);
 * @return {string[]} [intendedCommand, intendedTerms] Array destructed assignment is recommended (see example)
 */
export const getIntendedInput = (commands, command, terms = []) => {
    const VALID_COMMANDS = keys(commands);
    const {bestMatch: {target: intendedCommand}} = findBestMatch(command, VALID_COMMANDS);
    const VALID_TERMS = keys(commands[intendedCommand]);
    const intendedTerms = terms.map(term => findBestMatch(term, VALID_TERMS).bestMatch.target);
    return {intendedCommand, intendedTerms};
};
/**
 * Use npm CLI to return array of module versions
 * @param {string} name npm module name
 * @example
 * const versions = getVersions('react');
 * @return {string[]} Array of versions
 */
export const getVersions = async (name = '') => (name.length === 0) ? [] : (await execa('npm', ['view', name, 'versions']))
    .stdout
    .split(',\n')
    .map(str => str.match(/\d+[.]\d+[.]\d+/))
    .filter(isArray)
    .map(head)
    .map(semver.valid)
    .filter(Boolean);
/**
 * Install dependencies with npm
 * @param {string[]} [dependencies=[]] Modules to install
 * @param {Object} options Options to configure installation
 * @param {boolean} [options.dev=false] If true, add "--save-dev"
 * @param {boolean} [options.latest=true] If true, add "@latest" to all module names
 * @param {boolean} [options.skipInstall=false] Do not install (mostly for testing)
 * @example <caption>Install production dependencies</caption>
 * install(['react']);
 * @example <caption>Install development dependencies</caption>
 * install(['jest', 'babel-jest'], {dev: true});
 * @return {string[]} Array of inputs (mostly for testing)
 */
export const install = async (dependencies = [], options = {dev: false, latest: true, skipInstall: false}) => {
    const {dev, latest, skipInstall} = options;
    const identity = i => i;
    const concat = val => str => str + val;
    const args = ['install']
        .concat(dependencies
            .filter(name => validate(name).validForNewPackages)
            .map(latest ? concat('@latest') : identity)
        )
        .concat(dev ? '--save-dev' : []);
    skipInstall || await execa('npm', args);
    return args;
};
export const uninstall = async (dependencies = []) => {
    const args = ['uninstall'].concat(dependencies.filter(name => validate(name).validForNewPackages));
    (args.length === 1) || await execa('npm', args);
    return args;
};
/**
 * Add async tasks to a queue, handle completion with actions dispatched via dispatch function
 * @param {Object} data Data to be used for populating queue
 * @param {Queue} [data.queue={}] p-queue instance
 * @param {Object[]} [data.tasks=[]] Array of task objects
 * @param {function} [data.dispatch=()=>{}] Function to dispatch task completion (complete, skip, error) actions
 * @param {Object} [data.options={}] Options object to pass to task function
 * @return {undefined} Returns nothing (side effects only)
 */
export async function populateQueue(data = {queue: {}, tasks: [], dispatch: () => { }, options: {skipInstall: false}}) {
    const {queue, tasks, dispatch, options} = data;
    const {skipInstall} = options;
    const isNotOffline = skipInstall || await isOnline();
    const customOptions = assign({}, tasks.filter(complement(isValidTask)).reduce((acc, val) => assign(acc, val), options), {isNotOffline});
    dispatch({type: 'status', payload: {online: isNotOffline}});
    for (const [index, item] of tasks.filter(isValidTask).filter(isUniqueTask).entries()) {
        const {condition, task} = item;
        try {
            if (await condition(customOptions)) {
                await queue
                    .add(() => task(customOptions))
                    .then(() => dispatch({type: 'complete', payload: index}))
                    .catch(() => dispatch({
                        type: 'error', payload: {
                            index,
                            title: 'Failed to add task to queue',
                            location: 'task',
                            details: item.text
                        }
                    }));
            } else {
                dispatch({type: 'skipped', payload: index});
            }
        } catch (error) {
            dispatch({
                type: 'error',
                payload: {
                    error,
                    index,
                    title: 'Failed to test task conditions',
                    location: 'condition',
                    details: item.text
                }
            });
        }
    }
}
/**
 * Create and edit a Babel.js configuration file with a fluent API
 * @type {ModuleEditor}
 * @example <caption>Extend module.exports content and prepend text to the top of the file</caption>
 * await (new BabelConfigModuleEditor())
 *     .create()
 *     .extend({presets: [`'@babel/preset-env'`]})
 *     .prepend(`const {existsSync} = require('fs-extra');`)
 *     .commit();
 */
export const BabelConfigModuleEditor = createModuleEditor('babel.config.js', {
    plugins: [
        `'@babel/plugin-transform-runtime'`,
        `'@babel/plugin-proposal-class-properties'`,
        `'@babel/plugin-proposal-export-default-from'`,
        `'@babel/plugin-proposal-optional-chaining'`
    ],
    presets: [`'@babel/preset-env'`, `'babel-preset-minify'`]
});
/**
 * Create and edit an ESLint configuration file with a fluent API
 * @type {ModuleEditor}
 * @example
 * await (new EslintConfigModuleEditor())
 *     .create()
 *     .commit();
 */
export const EslintConfigModuleEditor = createModuleEditor('.eslintrc.js', {
    env: {
        es6: true,
        jest: true
    },
    extends: [
        `'omaha-prime-grade'`
    ],
    parser: `'babel-eslint'`
});
/**
 * Create and edit a package.json manifest file with a fluent API
 * @type {JsonEditor}
 * @example <caption>Create a new package.json</caption>
 * await (new PackageJsonEditor())
 *     .create()
 *     .commit();
 * @example <caption>Create a new package.json and read its contents (chaining OK)</caption>
 * const contents = (new PackageJsonEditor())
 *     .create()
 *     .read();
 * @example <caption>Extend a package.json</caption>
 * const script = {test: 'jest --coverage'};
 * await (new PackageJsonEditor())
 *     .extend({script})
 *     .commit();
 * @example <caption>Create and extend a package.json without writing to disk (chaining OK)</caption>
 * const script = {
 *     lint: 'eslint index.js -c ./.eslintrc.js'
 * };
 * await (new PackageJsonEditor())
 *     .create(false)
 *     .extend({script}, false)
 *     .commit();
 */
export const PackageJsonEditor = createJsonEditor('package.json', {
    name: 'my-project',
    version: '0.0.0',
    description: 'A super cool app/server/tool/library/widget/thingy',
    license: 'MIT',
    keywords: []
});
export const BsConfigJsonEditor = createJsonEditor('bsconfig.json', {
    'bs-dependencies': ['reason-react'],
    'bsc-flags': ['-bs-super-errors'],
    namespace: true,
    'package-specs': [{
        module: 'es6',
        'in-source': true
    }],
    'ppx-flags': [],
    reason: {'react-jsx': 3},
    refmt: 3,
    sources: [{
        dir: 'src',
        subdirs: true
    }],
    suffix: '.bs.js'
});
/**
 * Create and edit a PostCSS configuration file with a fluent API
 * @type {ModuleEditor}
 * @example
 * await (new PostcssConfigEditor())
 *     .create()
 *     .commit();
 */
export const PostcssConfigEditor = createModuleEditor('postcss.config.js', {
    map: true,
    parser: `require('postcss-safe-parser')`
});
/**
 * Create and edit a PurgeCSS configuration file with a fluent API
 * @type {ModuleEditor}
 * @example
 * await (new PurgecssConfigEditor())
 *     .create()
 *     .commit();
 */
export const PurgecssConfigEditor = createModuleEditor('purgecss.config.js', {
    content: [`'./assets/index.html'`]
});
/**
 * Create and edit a Rollup configuration file with a fluent API
 * @type {ModuleEditor}
 * @example
 * await (new RollupConfigEditor())
 *     .create()
 *     .commit();
 */
export const RollupConfigEditor = createModuleEditor('rollup.config.js', {
    input: `'./src/main.js'`,
    output: {
        file: `'./dist/bundle.min.js'`,
        format: `'iife'`,
        sourceMap: `'inline'`
    },
    plugins: [
        `babel({exclude: 'node_modules/**', runtimeHelpers: true})`,
        oneLineTrim`commonjs({
            namedExports: {
                './node_modules/backbone/backbone.js': ['Model', 'history'],
                './node_modules/backbone.marionette/lib/backbone.marionette.js': ['Application', 'View', 'MnObject']
            }
        })`,
        `resolve({browser: true})`,
        `replace({'process.env.NODE_ENV': JSON.stringify('production')})`,
        `terser()`
    ]
}, {esm: true});
/**
 * Create and edit a Webpack configuration file with a fluent API
 * @type {ModuleEditor}
 * @example
 * await (new WebpackConfigEditor())
 *     .create()
 *     .commit();
 */
export const WebpackConfigEditor = createFunctionModuleEditor('webpack.config.js', {
    mode: `'development'`,
    entry: {
        app: `'./src/main.js'`
    },
    output: {
        path: `resolve('./dist')`,
        filename: `'bundle.min.js'`
    },
    module: {
        rules: [
            {
                test: `/\.jsx?$/`,
                exclude: `/node_modules/`,
                loader: `'babel-loader'`,
                query: {
                    presets: [`'@babel/env'`]
                }
            }
        ]
    },
    plugins: [
        `new DashboardPlugin()`
    ]
}, {params: ['env', 'argv']});