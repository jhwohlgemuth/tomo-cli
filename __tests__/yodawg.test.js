import {existsSync} from 'fs';
import {join} from 'path';
import {
    removeAttributes,
    getDirectoryTree,
    useTemporaryDirectory
} from './tomo-test';
import {partialRight} from 'lodash';
import {format} from '../src/utils/common';

describe('Tomo testing tools', () => {
    test('can create and cleanup temporary directories', async () => {
        const [setTempDir, cleanupTempDir] = useTemporaryDirectory();
        const tempDir = await setTempDir();
        expect(existsSync(tempDir)).toBeTruthy();
        await cleanupTempDir();
        expect(existsSync(tempDir)).not.toBeTruthy();
    });
    test('can format objects for use as snapshots', () => {
        const removePaths = partialRight(removeAttributes, 'path');
        const result = removePaths({
            path: 'a',
            type: 'directory',
            children: [
                {
                    path: 'aa',
                    type: 'file'
                },
                {
                    path: 'b',
                    type: 'directory',
                    children: [
                        {
                            path: 'c',
                            type: 'directory',
                            children: [
                                {
                                    path: 'd',
                                    type: 'file'
                                }
                            ]
                        }
                    ]
                }
            ]
        });
        expect(format(result)).toMatchSnapshot();
    });
    test('get directory tree', () => {
        const folderpath = join(__dirname, 'tomo-fixtures/test-folder-tree');
        const tree = getDirectoryTree(folderpath);
        expect(tree).toMatchSnapshot();
        const augmentedTree = getDirectoryTree(folderpath, {omit: ['path', 'name']});
        expect(augmentedTree).toMatchSnapshot();
    });
});