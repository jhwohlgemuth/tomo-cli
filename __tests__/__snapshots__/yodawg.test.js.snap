// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tomo testing tools can format objects for use as snapshots 1`] = `
"{
    type: directory,
    children: [
        {
            type: file
        },
        {
            type: directory,
            children: [
                {
                    type: directory,
                    children: [
                        {
                            type: file
                        }
                    ]
                }
            ]
        }
    ]
}
"
`;

exports[`Tomo testing tools get directory tree 1`] = `
"{
    name: test-fold,
    children: [
        {
            name: a.txt,
            type: file
        },
        {
            name: b,
            children: [
                {
                    name: b.txt,
                    type: file
                }
            ],
            type: directory
        },
        {
            name: c,
            children: [
                {
                    name: c.txt,
                    type: file
                },
                {
                    name: d,
                    children: [
                        {
                            name: d.txt,
                            type: file
                        }
                    ],
                    type: directory
                }
            ],
            type: directory
        }
    ],
    type: directory
}
"
`;

exports[`Tomo testing tools get directory tree 2`] = `
"{
    children: [
        {
            extension: .txt,
            type: file
        },
        {
            children: [
                {
                    extension: .txt,
                    type: file
                }
            ],
            type: directory
        },
        {
            children: [
                {
                    extension: .txt,
                    type: file
                },
                {
                    children: [
                        {
                            extension: .txt,
                            type: file
                        }
                    ],
                    type: directory
                }
            ],
            type: directory
        }
    ],
    type: directory
}
"
`;
